### PR TITLE
Stop overcounting dead groups and increase your accuracy

### LINKED ISSUE #6367: Option to exclude zero-variance/dead groups from the loss denominator, accuracy increases
### Feature request

GRPO's per-group advantage is exactly `0` for every completion in a group where all K sampled completions got the same reward (common with binary/verifier rewards). But every current `loss_type` (`grpo`, `dapo`, `bnpo`, `dr_grpo`, `cispo`, `sapo`, `luspo`, `vespo`) still divides by a token/sequence count that includes those zero-advantage rows, diluting the average by whatever fraction of groups are dead that step.

We ran a fix and found immediate accuracy increase in all our experiments. Accuracy numbers on a public dataset  are as follows: `GSM8K with Qwen2.5-1.5B-Instruct (step 100, n=500, greedy eval)`

| Setup | Accuracy | Change |
|---|---|---|
| Binary verifier | 0.536 | baseline |
| Binary verifier, denominator fixed | 0.560 | +2.4pp, same reward |
| Dense reward model | 0.590 | +5.4pp total |

Fixing only the denominator (no reward change) recovered 2.4 points accuracy. A literature survey told us that this method is also part of broader set of methods proposed in AGPO (Adaptive Group Policy Optimization," arXiv:2503.15952).

**Disclosure:** this issue and the linked PR were prepared with AI-assisted research and drafting

### Motivation

We are throwing away accuracy by overcounting in denominator. Let's fix that.

### Your contribution

- PR for the said feature, will raise immediately after this issue
- uncovered a potential bug while working on this PR, filing as a separate issue and (maybe separate) PR

### PR DESCRIPTION
**Pull request by SushantDaga**
## What this PR do
[Closes #6367]

Adds one new opt-in value to `GRPOConfig.loss_type`: `"dapo_zv"`. Identical to `"dapo"` (the current default) in every way except the loss denominator excludes tokens belonging to zero-variance ("all completions in the group got the same reward") groups. Default (`"dapo"` and every other existing value) is completely unchanged.

## Why
Dividing by a token count that includes zero-signal rows silently shrinks the effective learning rate on the rows that DO carry signal, in proportion to how many groups are tied this step. We ran a fix and found immediate accuracy increase in all our experiments. Accuracy numbers on a public dataset  are as follows: `GSM8K with Qwen2.5-1.5B-Instruct (step 100, n=500, greedy eval)`

| Setup | Accuracy | Denominator |
|---|---|---|
| Plain binary (`loss_type="dapo"`) | 0.536 | all tokens, including dead groups |
| Binary + this fix (`loss_type="dapo_zv"`, equivalent) | 0.560 | dead-group tokens excluded |
| Dense reward model (`loss_type="dapo"`) | 0.590 | rarely hits zero-variance, so rarely diluted |

Fixing only the denominator (no reward change) recovered 2.4 points accuracy. A literature survey told us that this method is also part of broader set of methods proposed in AGPO (Adaptive Group Policy Optimization," arXiv:2503.15952).

## Relationship to prior TRL work on zero-std groups
- **#4060 (merged replay buffer)**: substitutes zero-std groups by iversampling. This is the orthogonal accounting remedy for runs not using it.
- **#5588 (closed not-a-bug) and its unmerged PRs #5597/#5640**: proposed masking the whole completion *including the KL term*; declined for changing loss values under `beta>0` and for departing from the canonical per-token-KL objective. This PR is scoped to avoid exactly those objections: **denominator only, KL/`beta` path untouched, opt-in `loss_type` value, no default changes** (the KL/beta term is now independently normalized specifically so this claim holds for the *loss value*, not just the KL computation, in every case, not only when there happen to be no dead groups.)

## Tests [AI generated]
Two layers, both passing:
- Standalone unit tests (pure `torch`, no GPU, no TRL import) — 10 tests covering the underlying arithmetic in isolation, including the default-unchanged property and an independently hand-computed expected-value check.
- `tests/test_grpo_trainer.py` — 11 dedicated tests plus 2 extended `loss_type` parametrizations (`test_train_loss_types[dapo_zv]`, `test_warning_raised_sequence_level_with_token_summed_loss[dapo_zv]`). Covers: bit-identical-to-`"dapo"` with no dead groups; exact-denominator-rescale with a mixed batch; the liger-kernel guard; `num_items_in_batch_zv` is tool_mask-aware and stays consistent under `mask_truncated_completions`; the normalizer clamp doesn't diverge from `"dapo"` for small-but-nonzero multi-process batches; NaN-in-group doesn't prevent detecting genuinely tied siblings; the all-dead-batch-with-KL-on case matches `"dapo"` exactly (no explosion); the KL/beta term isn't inflated by zero-variance exclusion in mixed batches either; and `num_items_in_batch_zv` is never computed for any other `loss_type` (no unconditional extra cost for non-`dapo_zv` users). Plus one guard test each in `test_grpo_with_replay_buffer_trainer.py`, `test_gspo_token_trainer.py`, and `test_gmpo_trainer.py` (`NotImplementedError` at construction), and two tests in `test_minillm_trainer.py` (raises with `rkl_advantage=True`, the default; does NOT raise with `rkl_advantage=False`, where `dapo_zv` behaves correctly).

## What still needs your call
1. Zero-variance status treated as fixed across `num_iterations` batch reuse (matching how `total_completion_tokens` is already treated)
2. Liger fused-loss path (`LigerFusedLinearGRPOLoss`): still not implemented; using `use_liger_kernel=True` with `loss_type="dapo_zv"` raises `NotImplementedError` at construction rather than silently mis-normalizing.

## Alternative API shape considered
This PR implements the fix as a new `loss_type` value (`"dapo_zv"`), matching the precedent CISPO (#4495) and LUSPO (#4988) both set for adding a loss variant to this same enum with motivation of small diff, narrow review surface, no interaction with any other option. An equally valid shape would be a boolean `GRPOConfig` flag (e.g. `exclude_zero_variance_from_loss`, default `False`) that composes with *every* existing `loss_type` rather than only `"dapo"`'s normalization family. Happy to reshape to the boolean-flag composition if maintainers prefer it. The new `loss_type` shape was chosen after reviewing CONTRIBUTING.md and how past PRs were accepted.

## Pre-empting "isn't this just `dr_grpo` / `bnpo` again?"
`dr_grpo` and `bnpo` change WHICH constant/count normalizes the loss (a global constant vs. local active-token count), orthogonal to whether zero-variance groups are included in that count. `"dapo_zv"` is `"dapo"`'s existing normalization scheme with dead-group tokens carved out of it.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure
- [ ] No AI usage: the PR was written entirely by a human.
- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

**Care is taken to keep PR simple, on the topic, with complete tests, and reference any closely related previous work including PR and issues**

## Who can review?
Anyone in the community is free to review the PR once the tests have passed.

---

> [!NOTE]
> **Medium Risk**
> Changes core GRPO loss normalization and KL/entropy scaling for an opt-in path; default `dapo` behavior is unchanged, but incorrect `num_items_in_batch_zv` logic would skew training gradients.
> 
> **Overview**
> Adds opt-in **`loss_type="dapo_zv"`** on **`GRPOConfig`** / **`GRPOTrainer`**: same policy term as **`dapo`**, but the loss mean uses **`num_items_in_batch_zv`** (active completion tokens from groups with non-identical rewards) instead of all tokens. **`num_items_in_batch_zv`** is computed only when this loss type is selected; zero-variance detection is NaN-safe and uses the same completion-length / **`tool_mask`** basis as **`dapo`**.
> 
> **KL (`beta`)**, **entropy bonus**, and the all–zero-variance batch case are handled so normalization matches **`dapo`** where intended (no inflated KL/entropy from the ZV denominator). **`use_liger_kernel=True`** with **`dapo_zv`** raises at construction.
> 
> Experimental trainers (**GMPO**, **gspo_token**, **GRPOWithReplayBuffer**) reject **`dapo_zv`** at init; **MiniLLM** rejects it with default **`rkl_advantage=True`** only. Docs add an AGPO paper index entry; tests cover equivalence, rescaling, edge cases, and guards.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 6d47c4094a53f8f95ace06b4e6ceaa494e42c823. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -2581,6 +2597,42 @@ def _generate_and_score_completions(
...
+            # "batch" it answers "is the WHOLE BATCH zero-variance" instead, silently degrading `dapo_zv` to plain
+            # `dapo`. Independent of `multi_objective_aggregation` too: under "normalize_then_sum", a group tied on
+            # raw `rewards` is tied at exactly 0 after its own per-function, per-group normalization (mean-
+            # centering a constant against itself is always 0), and since every function's within-group residuals
+            # are zero-mean, the final normalization step's batch-wide mean is also exactly 0 -- so the group's
+            # advantage is 0 there too, same as "sum_then_normalize"'s more direct group-mean-centering.
+            grouped_rewards = rewards.view(-1, num_generations)
+            group_max = torch.nan_to_num(grouped_rewards, nan=float("-inf")).amax(dim=1)
+            group_min = torch.nan_to_num(grouped_rewards, nan=float("inf")).amin(dim=1)
+            row_is_zero_variance = ((group_max - group_min).abs() <= 1e-6).repeat_interleave(num_generations)
```
[the remark this answers is not in the document]
Comment (SushantDaga) [reply trl/trainer/grpo_trainer.py:2620]: False positive. This is not wrong. The bot is doing local maths while ignoring what's happening overall. Adding test for it explicitly.  Every group independently sums to 0, the batch-wide mean subtracted afterward is also always ≈0 (average of zeros). for a genuinely tied group, the combined reward is forced to exactly 0 for every row, and subtracting the (also-0) batch mean changes nothing — advantage stays ≈0.

**Comment by qgallouedec** [comment]
Thanks @SushantDaga, this is careful, well-tested work and the writeup is thorough.

That said, I'm going to decline it on scope. The zero-variance-group question in the loss has already been through the repo: #5588 was closed not-planned, and #4060 (replay-buffer substitution of 0-std groups) is the remedy we merged for it, so a new opt-in `loss_type` for the same phenomenon adds API and maintenance surface (a variant threaded through the whole dapo path plus `NotImplementedError` guards in five trainers) for something we consider addressed. 

The evidence is also thin for a permanent public loss_type: one small self-reported eval (GSM8K / Qwen2.5-1.5B, n=500) drawn from a single paper. If you'd like to keep iterating, `trl/experimental/` is the right home for a speculative loss variant, and broader results across models/datasets would help make the case, happy to point you there.
